altos-rust/altos-core/src/sched/mod.rs

```diff
@@ -74,26 +80,52 @@ pub fn switch_context() {
         }
       }
 
-      'main: loop {
-        for i in Priority::all() {
-          while let Some(mut new_task) = PRIORITY_QUEUES[i].dequeue() {
-            if new_task.destroy {
-              drop(new_task);
-            }
-            else {
-              new_task.state = State::Running;
-              // UNSAFE: Accessing CURRENT_TASK
-              unsafe { CURRENT_TASK = Some(new_task) };
-              break 'main;
-            }
-          }
-        }
-      }
+      //NOTE May cause additional overhead by calling a stack frame
+      while schedule_next_task() { }
     },
     None => panic!("switch_context - current task doesn't exist!"),
   }
 }
 
+///Get the next task from the queue
+///If more than NORMAL_TASK_MAX Normal tasks have run don't try and schedule
+///a normal priorty task and give low priority a shot at running.
+///NOTE: Ranges could be replaced with constants to improve performace
+fn schedule_next_task() -> bool {
+  if NORMAL_TASK_COUNTER.load(Ordering::Relaxed)  >= NORMAL_TASK_MAX {
+    if !select_task(Priority::get_range_from(Priority::Critical, Priority::Normal)) {
+      return false;
+    }
+    NORMAL_TASK_COUNTER.store(0, Ordering::Relaxed);
+    return select_task(Priority::get_range_to_end_from(Priority::Low));
+  }
+  select_task(Priority::get_range_to_end_from(Priority::Critical))
+}
+
+
+//Select the next task from in priority the range 
+fn select_task(queues: ::core::ops::Range<usize>) -> bool {
```

Comment:
It doesn't seem intuitive that a function called select_task would return false when a task is selected for running and true when a task isn't selected.